nautobot/docs/development/apps/api/platform-features/table-extensions.md

```diff
@@ -0,0 +1,64 @@
+# Extending Tables
+
++++ 2.9.99-alpha
+
+Apps can extend any model-based `Table` classes that are provided by the Nautobot core.
+
+The requirements to extend a table are:
+
+* The file must be named `table_extensions.py`
+* The variable `table_extensions` must be declared in that file, and contain a list of `TableExtension` subclasses
+* The `model` attribute of each `TableExtension` subclass must be set to a valid model name in the dotted pair format (`{app_label}.{model}`, e.g. `tenant.tenant` or `dcim.device`)
+
+There are four ways that `TableExtension` can be used to modify a table.
+
+* Custom columns can be added to a table.
+* Core and custom columns can be added to the default visible columns.
+* Core columns can be removed from the default visible columns when they are not used or applicable.
```

Comment:
Interesting questions, I'll start with the last.

Because of the way this is implemented in the `CustomApp.NautobotAppConfig` class, it would be difficult to implement a weight system, because it would require knowledge of all modifications.  So it would need to be redesigned to accommodate that.   I can see the desire to have deterministic results.

There should be no issue if multiple apps remove columns from the same table. The `default_columns` is rebuilt using a comprehension that checks filters out any column names that are in the `remove_from_default_columns` list.  For example: When App 'A' is initialized it will remove column 'a' from the XYZTable, then when App 'B' is initialized column 'a' will no longer be in `default_columns`.  If App 'B' want to remove the same or different columns, it should not encounter any issues.

Last case: If one app wants to remove a column and another app want to add that same column.  The outcome would be determined based on the order the apps are initialized, with the last initialized app winning the battle.  I believe this scenario would not be common because:
1. If an app is adding a column to the table's `default_columns`, it is either a new column added by that app, and would be prefixed with the app_name.
2. If an app is adding a core column to the table's `default_columns`, it is likely not in the defaults and so it would not need to be removed.
3. If an app is removing a core column from the table's `default_columns`, it is likely already in the defaults and would not need to be added.